src/include/txn_inline.h

```diff
@@ -1040,9 +1045,25 @@ __wt_txn_read(
 
     /* If there's no visible update in the update chain or ondisk, check the history store file. */
     if (F_ISSET(S2C(session), WT_CONN_HS_OPEN) && !F_ISSET(session->dhandle, WT_DHANDLE_HS)) {
+        /*
+         * Generate a local key for the history store search from the cursor position slot when the
+         * caller doesn't pass it.
+         */
+        if (S2BT(session)->type == BTREE_ROW && key == NULL) {
+            WT_RET(__wt_scr_alloc(session, 0, &unpack_key));
+            WT_RET(__wt_row_leaf_key(
+              session, cbt->ref->page, &cbt->ref->page->pg_row[cbt->slot], unpack_key, true));
+            key = unpack_key;
+        }
         __wt_timing_stress(session, WT_TIMING_STRESS_HS_SEARCH);
-        WT_RET(__wt_hs_find_upd(session, S2BT(session)->id, key, cbt->iface.value_format, recno,
-          cbt->upd_value, &cbt->upd_value->buf));
+        ret = __wt_hs_find_upd(session, S2BT(session)->id, key, cbt->iface.value_format, recno,
+          cbt->upd_value, &cbt->upd_value->buf);
+        /* Free the locally allocated key. */
+        if (unpack_key != NULL) {
+            __wt_scr_free(session, &unpack_key);
+            key = NULL;
+        }
+        WT_RET(ret);
```

Comment:
Hari, if we always set `cbt.tmp` when `cbt.compare == 0`, then there's exactly one path to this code: we don't get an exact match in `__wt_btcur_search_near()`, which calls `__wt_cursor_valid()`, which calls `__wt_txn_read()`.

We could then move this code out of `__wt_txn_read()` and into `__wt_btcur_search_near()` (or `__wt_cursor_valid()` if you prefer?)

At the least, we can avoid the allocation of a scratch buffer in this path (use `cbt.tmp` instead), and it clarifies what's happening.

Reply:
Fixed in [ce8c9b7](https://github.com/wiredtiger/wiredtiger/pull/7777/commits/ce8c9b773c791ad6d224a58fa23b4015a2f6411f).